packages/experience-builder-sdk/src/core/editor/designComponentUtils.ts

```diff
@@ -0,0 +1,110 @@
+import { EntityStore } from '@contentful/visual-sdk';
+import {
+  CompositionNode,
+  CompositionDataSource,
+  CompositionUnboundValues,
+  CompositionComponentNode,
+  CompositionComponentPropValue,
+  Composition,
+} from '../../types';
+import { generateRandomId } from '../../utils';
+import { designComponentsRegistry } from '../../blocks/VisualEditorContext';
+
+export const deserializeDesignComponentNode = ({
+  node,
+  nodeId,
+  parentId,
+  experienceDataSource,
+  experienceUnboundValues,
+}: {
+  node: CompositionNode;
+  nodeId: string;
+  parentId?: string;
+  experienceDataSource: CompositionDataSource;
+  experienceUnboundValues: CompositionUnboundValues;
+}): CompositionComponentNode => {
+  const childNodeVariable: Record<string, CompositionComponentPropValue<'DesignValue'>> = {};
+  const dataSource: CompositionDataSource = {};
+  const unboundValues: CompositionUnboundValues = {};
+
+  for (const [variableName, variable] of Object.entries(node.variables)) {
+    if (variable.type === 'DesignValue') {
+      childNodeVariable[variableName] = variable;
+    } else if (variable.type === 'BoundValue') {
+      const [, uuid, ,] = variable.path.split('/');
+
+      dataSource[uuid] = { ...experienceDataSource[uuid] };
+    } else if (variable.type === 'UnboundValue') {
+      const uuid = variable.key;
+      unboundValues[uuid] = experienceUnboundValues[uuid];
+    }
+  }
+
+  const isDesignComponent = designComponentsRegistry.has(node.definitionId);
+
+  const children: CompositionComponentNode[] = node.children.map((child) =>
+    deserializeDesignComponentNode({
+      node: child,
+      nodeId: generateRandomId(16),
```

Comment:
will it be different each time? Should we try to use the existing id if it is already there?

Reply:
There's no existing node Id when we fetch the design component's children, which is why I generate a node id on deserialisation (similar to what we do in the web app)

Reply:
I thought that each node had a `data.id` in it (which is already a randomly generated string), so I meant that id

Reply:
The design component block nodes deserialized here in the SDK. The reason we have a node id normally is because the nodes are deserialized and sent from the web app. It's different for design components.

For design components, we get a design component id from the tree, fetch that entry, and deserialize the nodes here in the SDK to render. During the deserialization process here is when I'm generating a node id.